clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1197,14 +1181,23 @@ void handleResponse(AbstractResponse abstractResponse) {
                         continue;
                     }
                     boolean isInternal = cluster.internalTopics().contains(topicName);
-                    TreeMap<Integer, TopicPartitionInfo> partitions = new TreeMap<>();
                     List<PartitionInfo> partitionInfos = cluster.partitionsForTopic(topicName);
+                    List<TopicPartitionInfo> partitions = new ArrayList<>(partitionInfos.size());
                     for (PartitionInfo partitionInfo : partitionInfos) {
+                        Node leader = null;
+                        if (partitionInfo.leader() != null || partitionInfo.leader().id() != Node.noNode().id())
```

Comment:
Hmm, perhaps we should create a convenience static function for this, since it seems to crop up a lot?  Also, it looks like we should be use using the value of `leader` calculated here later in the function...